File: java/src/Solution0919.java
import dataStructure.TreeNode;

import java.util.LinkedList;
import java.util.Queue;

public class Solution0919 {
	public static void main(String[] args) {
//		TreeNode
//		CBTInserter cbtInserter = new CBTInserter();
	}
}

class CBTInserter {
	TreeNode root;
	Queue<TreeNode> queue;

	public CBTInserter(TreeNode root) {
		this.root = root;
		Queue<TreeNode> q = new LinkedList<>();
		q.offer(root);
		while (!q.isEmpty()) {
			TreeNode node = q.poll();
			if (node.left != null) {
				q.offer(node.left);
			}
			if (node.right != null) {
				q.offer(node.right);
			}
			if (node.left == null || node.right == null) {
				queue.offer(node);
			}
		}
	}

	public int insert(int val) {
		TreeNode node = new TreeNode(val);
		TreeNode t = queue.peek();
		int value = t.val;
		if (t.left == null) {
			t.left = node;
		} else if (t.right == null) {
			t.right = node;
			queue.poll();
		}
		queue.offer(node);
		return value;
	}

	public TreeNode get_root() {
		return root;
	}
}

File: GoLang/solution/solution0919.go
package solution

import datastructure "gproject/dataStructure"

type CBTInserter struct {
	root  *datastructure.TreeNode
	queue []*datastructure.TreeNode
}

func Constructor0919(root *datastructure.TreeNode) CBTInserter {
	queue := []*datastructure.TreeNode{}
	q := []*datastructure.TreeNode{root}
	for len(q) > 0 {
		node := q[0]
		q = q[1:]
		if node.Left != nil {
			q = append(q, node.Left)
		}
		if node.Right != nil {
			q = append(q, node.Right)
		}
		if node.Left == nil || node.Right == nil {
			queue = append(queue, node)
		}
	}
	return CBTInserter{root: root, queue: queue}
}

func (this *CBTInserter) Insert(val int) int {
	node := &datastructure.TreeNode{Val: val}
	t := this.queue[0]
	if t.Left == nil {
		this.queue[0].Left = node
	} else if t.Right == nil {
		this.queue[0].Right = node
		this.queue = this.queue[1:]
	}
	this.queue = append(this.queue, node)
	return t.Val
}

func (this *CBTInserter) Get_root() *datastructure.TreeNode {
	return this.root
}
